internals/daemon/api_notices.go

```diff
@@ -0,0 +1,154 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"regexp"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+)
+
+// A very loose regex to ensure client keys are in the form "domain.com/key"
+var clientKeyRegexp = regexp.MustCompile(`([a-z0-9-_]+\.)+[a-z0-9-_]+/.+`)
+
+func v1GetNotices(c *Command, r *http.Request, _ *UserState) Response {
+	query := r.URL.Query()
+
+	typeStr := query.Get("type")
+	noticeType := state.NoticeTypeFromString(typeStr)
+	if typeStr != "" && noticeType == "" {
+		return statusBadRequest("invalid notice type %q", typeStr)
+	}
+
+	key := query.Get("key")
+
+	afterStr := query.Get("after")
+	var after time.Time
+	if afterStr != "" {
+		var err error
+		after, err = time.Parse(time.RFC3339, afterStr)
+		if err != nil {
+			return statusBadRequest("invalid after timestamp %q: %v", afterStr, err)
+		}
+	}
+
+	filters := state.NoticeFilters{
+		Type:  noticeType,
+		Key:   key,
+		After: after,
+	}
+	var notices []*state.Notice
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	timeoutStr := query.Get("timeout")
+	if timeoutStr != "" {
+		// Wait up to timeout for notices matching given filters to occur
+		timeout, err := time.ParseDuration(timeoutStr)
+		if err != nil {
+			return statusBadRequest("invalid timeout %q: %v", timeoutStr, err)
+		}
+
+		ctx, cancel := context.WithTimeout(r.Context(), timeout)
+		defer cancel()
+
+		notices, err = st.WaitNotices(ctx, filters)
+		if errors.Is(err, context.DeadlineExceeded) {
+			return statusGatewayTimeout("timed out after %s", timeout)
+		}
+		if errors.Is(err, context.Canceled) {
+			return statusBadRequest("request canceled")
+		}
+		if err != nil {
+			return statusInternalError("cannot wait for notices: %s", err)
+		}
+	} else {
+		// No timeout given, fetch currently-available notices
+		notices = st.Notices(filters)
+	}
+
+	if len(notices) == 0 {
+		notices = []*state.Notice{} // avoid null result
+	}
+	return SyncResponse(notices)
+}
+
+func v1PostNotices(c *Command, r *http.Request, _ *UserState) Response {
+	var payload struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after"`
+		Data        map[string]string `json:"data"`
+	}
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&payload); err != nil {
+		return statusBadRequest("cannot decode request body: %v", err)
+	}
+
+	if payload.Action != "add" {
+		return statusBadRequest("invalid action %q", payload.Action)
+	}
+	if payload.Type != "client" {
+		return statusBadRequest(`invalid type %q (can only add "client" notices)`, payload.Type)
+	}
```

Comment:
Yeah, however, in case we have other actions later (or allow other types later), it's best to be explicit and require it now. This is similar to how we do it in other APIs like [`v1PostChange`](https://github.com/canonical/pebble/blob/3819000f4b2835484c1e981034987426efeeed85/internals/daemon/api_changes.go#L246-L248).

Reply:
my point is, `action` is already encoded in the HTTP verb. If the action was "change/modify", you should be using a PUT request, and for "remove", you should be using DELETE. What other actions would we need?

Adding fields to the API at a later stage should also be fine, since it is backward compatible and won't break the v1 semantic versioning. In that case, no need to put these in before we actually need them IMO.

Reply:
Once again, the precedent is strong with this one. I had this discussion with Gustavo a while ago, and Pebble's (and snap's) API are "vaguely REST-like if you squint" rather than REST-ful (using specific HTTP verbs for different actions and so on). Reads are done using GET and writes using POST, and we use meaningful URLs, but that's about it.

We use this "action" pattern for several other APIs in Pebble:

```
METH /path                actions
---------------------------------
POST /v1/changes/{id}     abort
GET  /v1/files            list read
POST /v1/files            make-dirs remove write
POST /v1/layers           add
POST /v1/services         autostart replan restart start stop
POST /v1/warnings         okay
```

So again, I'd much rather be consistent with that than break the pattern for the notices API.

As far as requiring the "action" field: I think you're right that it would be backwards-compatible to add it later if we needed it, and have the default action be "add" at that point. However, requiring it up-front is consistent with how we do it in the existing endpoints, and in light of the above API design, I think it's forward-looking to require it from the beginning. It certainly doesn't hurt to make things explicit.